### PR TITLE
Enable fseek for files in S3 storage

### PR DESCRIPTION
**Pull request by icewind1991**
Based on the work of @ahti in https://github.com/nextcloud/server/pull/19073

> This allows nextcloud to use fseek when using a S3 storage backend, which enables range requests and thus video streaming including fast seeking, and other applications.
> 
> #8275 should be fixed by this.
> 
> The implementation is a new stream wrapper around the normal http stream. When fseek is used, a new stream is opened with a range header to begin at the right offset.
> 
> I'm not really a PHP developer, so there might be bugs, and there is no proper error handling when the storage backend doesn't reply with a content-range header (although I would expect S3 backends to support range requests properly), and probably in some other places, too.
> 
> I have not investigated this, but maybe this approach could also work for the other object store backends?

Adjusted to be generic of any http stream where the server supports range requests so it's easy to extend to other storage backends in the future

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]
:elephant:

File: lib/private/Files/Stream/SeekableHttpStream.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+	private static $registered = false;
+
+	/**
+	 * Registers the stream wrapper using the `httpseek://` url scheme
+	 * $return void
+	 */
+	private static function registerIfNeeded() {
+		if (!self::$registered) {
+			stream_wrapper_register(
```
Comment (kesselb) [inline lib/private/Files/Stream/SeekableHttpStream.php]: > stream_wrapper_register() will return FALSE if the protocol already has a handler. 

https://www.php.net/manual/en/function.stream-wrapper-register.php

Do we really need a state (registered y/no)? I guess `stream_wrapper_register` will not register more than one handler for one protocol.

**Review by rullzer — APPROVED** [review]

**Comment by rullzer** [comment]
Needs a rebase

File: lib/private/Files/Stream/SeekableHttpStream.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		if ($begin !== $start) {
+			return false;
+		}
+
+		$this->offset = $begin;
+
+		return true;
+	}
+
+	function stream_open($path, $mode, $options, &$opened_path) {
```
Comment (kesselb) [inline lib/private/Files/Stream/SeekableHttpStream.php:113]:
```suggestion
	public function stream_open($path, $mode, $options, &$opened_path) {
```

File: lib/private/Files/Stream/SeekableHttpStream.php [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		$options = stream_context_get_options($this->context)[self::PROTOCOL];
+		$this->openCallback = $options['callback'];
+
+		return $this->reconnect(0);
+	}
+
+	function stream_read($count) {
```
Comment (kesselb) [inline lib/private/Files/Stream/SeekableHttpStream.php:120]:
```suggestion
	public function stream_read($count) {
```

File: lib/private/Files/Stream/SeekableHttpStream.php [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		if (!$this->current) {
+			return false;
+		}
+		$ret = fread($this->current, $count);
+		$this->offset += strlen($ret);
+		return $ret;
+	}
+
+	function stream_seek($offset, $whence = SEEK_SET) {
```
Comment (kesselb) [inline lib/private/Files/Stream/SeekableHttpStream.php:129]:
```suggestion
	public function stream_seek($offset, $whence = SEEK_SET) {
```

**Comment by rullzer** [comment]
@icewind1991 please address the comments and then lets get this in

**Comment by MorrisJobke** [comment]
@rullzer @icewind1991 Do we want to backport this to 17 and 18?

**Pull request merged**